### PR TITLE
feat: amplify upgrade and amplify uninstall commands

### PR DESCRIPTION
**Pull request by edwardfoyle**
Implements `amplify upgrade` and `amplify uninstall` commands.

When the CLI is running using node, upgrade and uninstall print messages telling the customer to upgrade / uninstall using npm

When the CLI is packaged:
- UPGRADE on mac / linux will check for a new binary and replace the existing one
- UPGRADE on windows will move the current installation to `amplify-old.exe` and download the new binary to `amplify.exe` (on windows the currently executing file is locked but can be renamed). Then on the next run of amplify (any command), it will delete `amplify-old.exe`
- UNINSTALL on mac / linux will remove the .amplify folder (which contains the binary)
- UNINSTALL on windows will move `amplify.exe` to `.amplify-pending-delete.exe` outside of the .amplify folder. It will then mark `.amplify-pending-delete.exe` for deletion using a registry value (indicates that the file should be deleted the next time the system is rebooted) and then delete the .amplify folder

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/commands/upgrade.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+const latestVersionUrl = 'https://api.github.com/repos/aws-amplify/amplify-cli/releases/latest';
+
+export const run = async (context: $TSContext) => {
+  const { version: thisVersion } = require('../../package.json');
+  if (typeof thisVersion !== 'string') {
+    throw new Error('Cannot determine current CLI version. Try uninstalling and reinstalling the CLI.');
+  }
+  const latestVersion = await getLatestVersion();
+  if (gt(latestVersion, thisVersion)) {
+    await upgradeCli(latestVersion);
```
Comment (jhockett) [inline packages/amplify-cli/src/commands/upgrade.ts:21]: nit: catch and handle errors that can get thrown by `updagePackagedCli`

**Review by jhockett — APPROVED** [review]
LGTM, just one nit

File: packages/amplify-cli/src/index.ts [hunk 2]
```diff
@@ -17,13 +17,14 @@ import { notify } from './version-notifier';
...
 import { ensureMobileHubCommandCompatibility } from './utils/mobilehub-support';
-import { postInstallInitialization } from './utils/post-install-initialization';
+import { deleteOldVersion } from './utils/win-utils';
 EventEmitter.defaultMaxListeners = 1000;
 
 // entry from commandline
 export async function run() {
   let errorHandler = (e: Error) => {};
   try {
+    deleteOldVersion();
```
Comment (SwaySway) [inline packages/amplify-cli/src/index.ts:27]: Is there a reason we always choose to delete the old version when running the CLI?

File: packages/amplify-cli/src/commands/upgrade.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+const upgradeCli = async (version: string) => {
+  const isWin = process.platform.startsWith('win');
+  const binPath = path.join(pathManager.getHomeDotAmplifyDirPath(), 'bin', isWin ? 'amplify.exe' : 'amplify');
+  const platformSuffix = isWin ? 'win.exe' : process.platform === 'darwin' ? 'macos' : 'linux';
+  const url = binUrl(version, platformSuffix);
+  const spinner = ora();
+
+  if (isWin) {
+    await fs.move(binPath, oldVersionPath);
```
Comment (SwaySway) [inline packages/amplify-cli/src/commands/upgrade.ts:43]: Is there a flow in which we need to keep the old ver on win? It looks like we delete it every time the CLI is executed.

File: packages/amplify-cli/src/index.ts (hunk 2 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/index.ts:27]: it only needs to run on the command after upgrade, but to know that we would need to maintain some state about what the last command was. it's easier to just call this method every time and if the old version isn't there it's a noop

File: packages/amplify-cli/src/commands/upgrade.ts (hunk 3 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/commands/upgrade.ts:43]: No we don't need to keep the old version

**Review by attilah — APPROVED** [review]
LGTM, if we could get rid of `hidefile` by using node-ffi-lib and the API I suggested and also `MoveFileEx` that could remove all the registry, hex magic we have in the PR.

File: packages/amplify-cli/src/commands/upgrade.ts [hunk 4]
```diff
@@ -25,32 +29,51 @@ export const run = async (context: $TSContext) => {
...
-
-  spinner.start('Downloading latest Amplify CLI version...');
   const response = await fetch(url);
   if (response.status >= 400) {
     throw new Error(`${response.status}: Request to ${url} failed:\n${JSON.stringify(response.json(), null, 2)}`);
   }
-  const bin = await response.buffer();
-  spinner.succeed('Download complete!');
-  await fs.writeFile(binPath, bin);
+  const len = response.headers.get('content-length');
```
Comment (kaustavghosh06) [inline packages/amplify-cli/src/commands/upgrade.ts:55]: content-length is always guaranteed in http header right?

**Review by kaustavghosh06 — APPROVED** [review]
LGTM. - just reviewd the last commit 8265995aa1901b3ed7056564e03c668c89729a62 as per request.

File: packages/amplify-cli/src/commands/upgrade.ts (hunk 4 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/commands/upgrade.ts:55]: not always, but GH download APIs include it.

**Pull request merged**